File: 05_arrays/05_13_two_sum/TwoSum.java
import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    public static void main(String[] args) {
        int[] arr = {2, 6, 5, 8, 11};
        int target = 14;

        System.out.println("Elements:-");
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();

        System.out.println("Target: " + target);

        int[] result = twoSum(arr, target);
        System.out.println("This is the answer for variant 2: [" + result[0] + ", " + result[1] + "]");
    }
    
    private static int[] twoSum(int[] arr, int target) {
        Map<Integer, Integer> numMap = new HashMap<>();

        for (int i = 0; i < arr.length; i++) {
            int complement = target - arr[i];
            if (numMap.containsKey(complement)) {
                return new int[]{numMap.get(complement), i};
            }
            numMap.put(arr[i], i);
        }
        
        return new int[] { -1, -1 };
    }
}

File: 05_arrays/05_13_two_sum/two_sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

vector<int> twoSum(vector<int> &nums, int target)
{
    unordered_map<int, int> uMap;

    for (int i = 0; i < nums.size(); i++)
    {
        int complement = target - nums[i];
        if (uMap.find(complement) != uMap.end())
        {
            return {uMap[complement], i};
        }
        uMap[nums[i]] = i;
    }

    return {-1, -1};
}

int main()
{
    int n = 5;
    vector<int> arr = {2, 6, 5, 8, 11};

    cout << "Elements:- " << endl;
    for (size_t i = 0; i < arr.size(); i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
    
    int target = 14;
    cout << "Target: " << target << endl;

    vector<int> result = twoSum(arr, target);
    cout << "This is the answer for variant 2: [" << result[0] << ", " << result[1] << "]" << endl;
    return 0;
}